File: open_source_projects/Deepbox-main/benchmarks/python/05_ml.py
"""
Benchmark 05 — Machine Learning
scikit-learn
"""

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

import numpy as np
from sklearn.linear_model import LinearRegression, Ridge, Lasso, LogisticRegression, SGDClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.svm import LinearSVC, LinearSVR
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor, GradientBoostingClassifier, GradientBoostingRegressor
from sklearn.cluster import KMeans, DBSCAN
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from utils import run, create_suite, header, footer

suite = create_suite("ml", "scikit-learn")
header("Benchmark 05 — Machine Learning", "scikit-learn")

# ── Data generators ──────────────────────────────────────

rng = np.random.RandomState(42)

def make_reg(n, f, seed=42):
    r = np.random.RandomState(seed)
    X = r.rand(n, f) * 10
    w = np.arange(1, f + 1, dtype=float)
    y = X @ w + (r.rand(n) - 0.5) * 2
    return X, y

def make_cls(n, f, seed=42):
    r = np.random.RandomState(seed)
    X = r.rand(n, f) * 10
    y = (X.sum(axis=1) > f * 5).astype(int)
    return X, y

def make_clust(n, f, seed=42):
    r = np.random.RandomState(seed)
    X = np.vstack([r.rand(n // 3, f) * 2 + i * 5 for i in range(3)])
    return X[:n]

Xr200, yr200 = make_reg(200, 5)
Xr500, yr500 = make_reg(500, 10)
Xc200, yc200 = make_cls(200, 5)
Xc500, yc500 = make_cls(500, 10)
Xcl200 = make_clust(200, 5)
Xcl500 = make_clust(500, 5)

# ── Linear Regression ───────────────────────────────────

run(suite, "LinearRegression fit", "200x5", lambda: LinearRegression().fit(Xr200, yr200))
run(suite, "LinearRegression fit", "500x10", lambda: LinearRegression().fit(Xr500, yr500))
lr = LinearRegression().fit(Xr200, yr200)
run(suite, "LinearRegression predict", "200x5", lambda: lr.predict(Xr200))

# ── Ridge ───────────────────────────────────────────────

run(suite, "Ridge fit", "200x5", lambda: Ridge(alpha=1.0).fit(Xr200, yr200))
run(suite, "Ridge fit", "500x10", lambda: Ridge(alpha=1.0).fit(Xr500, yr500))
ridge = Ridge(alpha=1.0).fit(Xr200, yr200)
run(suite, "Ridge predict", "200x5", lambda: ridge.predict(Xr200))

# ── Lasso ───────────────────────────────────────────────

run(suite, "Lasso fit", "200x5", lambda: Lasso(alpha=0.1).fit(Xr200, yr200))
run(suite, "Lasso fit", "500x10", lambda: Lasso(alpha=0.1).fit(Xr500, yr500))
lasso = Lasso(alpha=0.1).fit(Xr200, yr200)
run(suite, "Lasso predict", "200x5", lambda: lasso.predict(Xr200))

# ── Logistic Regression ─────────────────────────────────

run(suite, "LogisticRegression fit", "200x5", lambda: LogisticRegression(max_iter=200).fit(Xc200, yc200))
run(suite, "LogisticRegression fit", "500x10", lambda: LogisticRegression(max_iter=200).fit(Xc500, yc500))
logreg = LogisticRegression(max_iter=200).fit(Xc200, yc200)
run(suite, "LogisticRegression predict", "200x5", lambda: logreg.predict(Xc200))

# ── GaussianNB ──────────────────────────────────────────

run(suite, "GaussianNB fit", "200x5", lambda: GaussianNB().fit(Xc200, yc200))
run(suite, "GaussianNB fit", "500x10", lambda: GaussianNB().fit(Xc500, yc500))
gnb = GaussianNB().fit(Xc200, yc200)
run(suite, "GaussianNB predict", "200x5", lambda: gnb.predict(Xc200))

# ── KNN Classifier ──────────────────────────────────────

run(suite, "KNeighborsClassifier fit", "200x5", lambda: KNeighborsClassifier(n_neighbors=5).fit(Xc200, yc200))
run(suite, "KNeighborsClassifier fit", "500x10", lambda: KNeighborsClassifier(n_neighbors=5).fit(Xc500, yc500))
knnc = KNeighborsClassifier(n_neighbors=5).fit(Xc200, yc200)
run(suite, "KNeighborsClassifier predict", "200x5", lambda: knnc.predict(Xc200))

# ── KNN Regressor ───────────────────────────────────────

run(suite, "KNeighborsRegressor fit", "200x5", lambda: KNeighborsRegressor(n_neighbors=5).fit(Xr200, yr200))
knnr = KNeighborsRegressor(n_neighbors=5).fit(Xr200, yr200)
run(suite, "KNeighborsRegressor predict", "200x5", lambda: knnr.predict(Xr200))

# ── LinearSVC ───────────────────────────────────────────

run(suite, "LinearSVC fit", "200x5", lambda: LinearSVC(max_iter=200, dual=True).fit(Xc200, yc200))
run(suite, "LinearSVC fit", "500x10", lambda: LinearSVC(max_iter=200, dual=True).fit(Xc500, yc500))
svc = LinearSVC(max_iter=200, dual=True).fit(Xc200, yc200)
run(suite, "LinearSVC predict", "200x5", lambda: svc.predict(Xc200))

# ── LinearSVR ───────────────────────────────────────────

run(suite, "LinearSVR fit", "200x5", lambda: LinearSVR(max_iter=200, dual=True).fit(Xr200, yr200))
svr = LinearSVR(max_iter=200, dual=True).fit(Xr200, yr200)
run(suite, "LinearSVR predict", "200x5", lambda: svr.predict(Xr200))

# ── Decision Tree ───────────────────────────────────────

run(suite, "DecisionTreeClassifier fit", "200x5", lambda: DecisionTreeClassifier(max_depth=5).fit(Xc200, yc200))
run(suite, "DecisionTreeClassifier fit", "500x10", lambda: DecisionTreeClassifier(max_depth=5).fit(Xc500, yc500))
dtc = DecisionTreeClassifier(max_depth=5).fit(Xc200, yc200)
run(suite, "DecisionTreeClassifier predict", "200x5", lambda: dtc.predict(Xc200))

run(suite, "DecisionTreeRegressor fit", "200x5", lambda: DecisionTreeRegressor(max_depth=5).fit(Xr200, yr200))
dtr = DecisionTreeRegressor(max_depth=5).fit(Xr200, yr200)
run(suite, "DecisionTreeRegressor predict", "200x5", lambda: dtr.predict(Xr200))

# ── Random Forest ───────────────────────────────────────

run(suite, "RandomForestClassifier fit", "200x5", lambda: RandomForestClassifier(n_estimators=10, max_depth=5).fit(Xc200, yc200), iterations=5)
run(suite, "RandomForestClassifier fit", "500x10", lambda: RandomForestClassifier(n_estimators=10, max_depth=5).fit(Xc500, yc500), iterations=5)
rfc = RandomForestClassifier(n_estimators=10, max_depth=5).fit(Xc200, yc200)
run(suite, "RandomForestClassifier predict", "200x5", lambda: rfc.predict(Xc200))

run(suite, "RandomForestRegressor fit", "200x5", lambda: RandomForestRegressor(n_estimators=10, max_depth=5).fit(Xr200, yr200), iterations=5)
rfr = RandomForestRegressor(n_estimators=10, max_depth=5).fit(Xr200, yr200)
run(suite, "RandomForestRegressor predict", "200x5", lambda: rfr.predict(Xr200))

# ── Gradient Boosting ───────────────────────────────────

run(suite, "GradientBoostingClassifier fit", "200x5", lambda: GradientBoostingClassifier(n_estimators=10, max_depth=3).fit(Xc200, yc200), iterations=5)
gbc = GradientBoostingClassifier(n_estimators=10, max_depth=3).fit(Xc200, yc200)
run(suite, "GradientBoostingClassifier predict", "200x5", lambda: gbc.predict(Xc200))

run(suite, "GradientBoostingRegressor fit", "200x5", lambda: GradientBoostingRegressor(n_estimators=10, max_depth=3).fit(Xr200, yr200), iterations=5)
gbr = GradientBoostingRegressor(n_estimators=10, max_depth=3).fit(Xr200, yr200)
run(suite, "GradientBoostingRegressor predict", "200x5", lambda: gbr.predict(Xr200))

# ── KMeans ──────────────────────────────────────────────

run(suite, "KMeans fit", "200x5 k=3", lambda: KMeans(n_clusters=3, max_iter=50, n_init=1).fit(Xcl200))
run(suite, "KMeans fit", "500x5 k=3", lambda: KMeans(n_clusters=3, max_iter=50, n_init=1).fit(Xcl500))
km = KMeans(n_clusters=3, max_iter=50, n_init=1).fit(Xcl200)
run(suite, "KMeans predict", "200x5", lambda: km.predict(Xcl200))

# ── DBSCAN ──────────────────────────────────────────────

run(suite, "DBSCAN fit", "200x5", lambda: DBSCAN(eps=2.0, min_samples=5).fit(Xcl200))
run(suite, "DBSCAN fit", "500x5", lambda: DBSCAN(eps=2.0, min_samples=5).fit(Xcl500))

# ── PCA ─────────────────────────────────────────────────

run(suite, "PCA fit", "200x5 k=2", lambda: PCA(n_components=2).fit(Xcl200))
run(suite, "PCA fit", "500x5 k=3", lambda: PCA(n_components=3).fit(Xcl500))
pca = PCA(n_components=2).fit(Xcl200)
run(suite, "PCA transform", "200x5", lambda: pca.transform(Xcl200))

# ── TSNE ────────────────────────────────────────────────

run(suite, "TSNE fit", "200x5", lambda: TSNE(n_components=2, perplexity=30).fit_transform(Xcl200), iterations=3, warmup=1)

footer(suite, "sklearn-ml.json")

File: open_source_projects/Deepbox-main/benchmarks/deepbox/05-ml.ts
/**
 * Benchmark 05 — Machine Learning
 * Deepbox vs scikit-learn
 */

import {
	DBSCAN,
	DecisionTreeClassifier,
	DecisionTreeRegressor,
	GaussianNB,
	GradientBoostingClassifier,
	GradientBoostingRegressor,
	KMeans,
	KNeighborsClassifier,
	KNeighborsRegressor,
	Lasso,
	LinearRegression,
	LinearSVC,
	LinearSVR,
	LogisticRegression,
	PCA,
	RandomForestClassifier,
	RandomForestRegressor,
	Ridge,
	TSNE,
} from "deepbox/ml";
import { tensor } from "deepbox/ndarray";
import { createSuite, footer, header, run } from "../utils";

const suite = createSuite("ml");
header("Benchmark 05 — Machine Learning");

// ── Data generators ──────────────────────────────────────

function seededRng(seed: number) {
	let s = seed >>> 0;
	return () => {
		s = (s * 1664525 + 1013904223) >>> 0;
		return s / 2 ** 32;
	};
}

function makeRegData(n: number, f: number, seed: number) {
	const rand = seededRng(seed);
	const X: number[][] = [],
		y: number[] = [];
	for (let i = 0; i < n; i++) {
		const row: number[] = [];
		let v = 0;
		for (let j = 0; j < f; j++) {
			const x = rand() * 10;
			row.push(x);
			v += x * (j + 1);
		}
		X.push(row);
		y.push(v + (rand() - 0.5) * 2);
	}
	return { X: tensor(X), y: tensor(y) };
}

function makeClsData(n: number, f: number, seed: number) {
	const rand = seededRng(seed);
	const X: number[][] = [],
		y: number[] = [];
	for (let i = 0; i < n; i++) {
		const row: number[] = [];
		let s = 0;
		for (let j = 0; j < f; j++) {
			const x = rand() * 10;
			row.push(x);
			s += x;
		}
		X.push(row);
		y.push(s > f * 5 ? 1 : 0);
	}
	return { X: tensor(X), y: tensor(y) };
}

function makeClustData(n: number, f: number, seed: number) {
	const rand = seededRng(seed);
	const X: number[][] = [];
	for (let i = 0; i < n; i++) {
		const row: number[] = [];
		const c = i % 3;
		for (let j = 0; j < f; j++) row.push(c * 5 + rand() * 2);
		X.push(row);
	}
	return tensor(X);
}

const reg200 = makeRegData(200, 5, 42);
const reg500 = makeRegData(500, 10, 42);
const cls200 = makeClsData(200, 5, 42);
const cls500 = makeClsData(500, 10, 42);
const clust200 = makeClustData(200, 5, 42);
const clust500 = makeClustData(500, 5, 42);

// ── Linear Regression ───────────────────────────────────

run(suite, "LinearRegression fit", "200x5", () => new LinearRegression().fit(reg200.X, reg200.y));
run(suite, "LinearRegression fit", "500x10", () => new LinearRegression().fit(reg500.X, reg500.y));
const lr = new LinearRegression().fit(reg200.X, reg200.y);
run(suite, "LinearRegression predict", "200x5", () => lr.predict(reg200.X));

// ── Ridge ───────────────────────────────────────────────

run(suite, "Ridge fit", "200x5", () => new Ridge({ alpha: 1.0 }).fit(reg200.X, reg200.y));
run(suite, "Ridge fit", "500x10", () => new Ridge({ alpha: 1.0 }).fit(reg500.X, reg500.y));
const ridge = new Ridge({ alpha: 1.0 }).fit(reg200.X, reg200.y);
run(suite, "Ridge predict", "200x5", () => ridge.predict(reg200.X));

// ── Lasso ───────────────────────────────────────────────

run(suite, "Lasso fit", "200x5", () => new Lasso({ alpha: 0.1 }).fit(reg200.X, reg200.y));
run(suite, "Lasso fit", "500x10", () => new Lasso({ alpha: 0.1 }).fit(reg500.X, reg500.y));
const lasso = new Lasso({ alpha: 0.1 }).fit(reg200.X, reg200.y);
run(suite, "Lasso predict", "200x5", () => lasso.predict(reg200.X));

// ── Logistic Regression ─────────────────────────────────

run(suite, "LogisticRegression fit", "200x5", () =>
	new LogisticRegression().fit(cls200.X, cls200.y)
);
run(suite, "LogisticRegression fit", "500x10", () =>
	new LogisticRegression().fit(cls500.X, cls500.y)
);
const logReg = new LogisticRegression().fit(cls200.X, cls200.y);
run(suite, "LogisticRegression predict", "200x5", () => logReg.predict(cls200.X));

// ── GaussianNB ──────────────────────────────────────────

run(suite, "GaussianNB fit", "200x5", () => new GaussianNB().fit(cls200.X, cls200.y));
run(suite, "GaussianNB fit", "500x10", () => new GaussianNB().fit(cls500.X, cls500.y));
const gnb = new GaussianNB().fit(cls200.X, cls200.y);
run(suite, "GaussianNB predict", "200x5", () => gnb.predict(cls200.X));

// ── KNN Classifier ──────────────────────────────────────

run(suite, "KNeighborsClassifier fit", "200x5", () =>
	new KNeighborsClassifier({ nNeighbors: 5 }).fit(cls200.X, cls200.y)
);
run(suite, "KNeighborsClassifier fit", "500x10", () =>
	new KNeighborsClassifier({ nNeighbors: 5 }).fit(cls500.X, cls500.y)
);
const knnc = new KNeighborsClassifier({ nNeighbors: 5 }).fit(cls200.X, cls200.y);
run(suite, "KNeighborsClassifier predict", "200x5", () => knnc.predict(cls200.X));

// ── KNN Regressor ───────────────────────────────────────

run(suite, "KNeighborsRegressor fit", "200x5", () =>
	new KNeighborsRegressor({ nNeighbors: 5 }).fit(reg200.X, reg200.y)
);
const knnr = new KNeighborsRegressor({ nNeighbors: 5 }).fit(reg200.X, reg200.y);
run(suite, "KNeighborsRegressor predict", "200x5", () => knnr.predict(reg200.X));

// ── LinearSVC ───────────────────────────────────────────

run(suite, "LinearSVC fit", "200x5", () => new LinearSVC().fit(cls200.X, cls200.y));
run(suite, "LinearSVC fit", "500x10", () => new LinearSVC().fit(cls500.X, cls500.y));
const svc = new LinearSVC().fit(cls200.X, cls200.y);
run(suite, "LinearSVC predict", "200x5", () => svc.predict(cls200.X));

// ── LinearSVR ───────────────────────────────────────────

run(suite, "LinearSVR fit", "200x5", () => new LinearSVR().fit(reg200.X, reg200.y));
const svr = new LinearSVR().fit(reg200.X, reg200.y);
run(suite, "LinearSVR predict", "200x5", () => svr.predict(reg200.X));

// ── Decision Tree ───────────────────────────────────────

run(suite, "DecisionTreeClassifier fit", "200x5", () =>
	new DecisionTreeClassifier({ maxDepth: 5 }).fit(cls200.X, cls200.y)
);
run(suite, "DecisionTreeClassifier fit", "500x10", () =>
	new DecisionTreeClassifier({ maxDepth: 5 }).fit(cls500.X, cls500.y)
);
const dtc = new DecisionTreeClassifier({ maxDepth: 5 }).fit(cls200.X, cls200.y);
run(suite, "DecisionTreeClassifier predict", "200x5", () => dtc.predict(cls200.X));

run(suite, "DecisionTreeRegressor fit", "200x5", () =>
	new DecisionTreeRegressor({ maxDepth: 5 }).fit(reg200.X, reg200.y)
);
const dtr = new DecisionTreeRegressor({ maxDepth: 5 }).fit(reg200.X, reg200.y);
run(suite, "DecisionTreeRegressor predict", "200x5", () => dtr.predict(reg200.X));

// ── Random Forest ───────────────────────────────────────

run(
	suite,
	"RandomForestClassifier fit",
	"200x5",
	() => new RandomForestClassifier({ nEstimators: 10, maxDepth: 5 }).fit(cls200.X, cls200.y),
	{ iterations: 5 }
);
run(
	suite,
	"RandomForestClassifier fit",
	"500x10",
	() => new RandomForestClassifier({ nEstimators: 10, maxDepth: 5 }).fit(cls500.X, cls500.y),
	{ iterations: 5 }
);
const rfc = new RandomForestClassifier({ nEstimators: 10, maxDepth: 5 }).fit(cls200.X, cls200.y);
run(suite, "RandomForestClassifier predict", "200x5", () => rfc.predict(cls200.X));

run(
	suite,
	"RandomForestRegressor fit",
	"200x5",
	() => new RandomForestRegressor({ nEstimators: 10, maxDepth: 5 }).fit(reg200.X, reg200.y),
	{ iterations: 5 }
);
const rfr = new RandomForestRegressor({ nEstimators: 10, maxDepth: 5 }).fit(reg200.X, reg200.y);
run(suite, "RandomForestRegressor predict", "200x5", () => rfr.predict(reg200.X));

// ── Gradient Boosting ───────────────────────────────────

run(
	suite,
	"GradientBoostingClassifier fit",
	"200x5",
	() => new GradientBoostingClassifier({ nEstimators: 10, maxDepth: 3 }).fit(cls200.X, cls200.y),
	{ iterations: 5 }
);
const gbc = new GradientBoostingClassifier({
	nEstimators: 10,
	maxDepth: 3,
}).fit(cls200.X, cls200.y);
run(suite, "GradientBoostingClassifier predict", "200x5", () => gbc.predict(cls200.X));

run(
	suite,
	"GradientBoostingRegressor fit",
	"200x5",
	() => new GradientBoostingRegressor({ nEstimators: 10, maxDepth: 3 }).fit(reg200.X, reg200.y),
	{ iterations: 5 }
);
const gbr = new GradientBoostingRegressor({ nEstimators: 10, maxDepth: 3 }).fit(reg200.X, reg200.y);
run(suite, "GradientBoostingRegressor predict", "200x5", () => gbr.predict(reg200.X));

// ── KMeans ──────────────────────────────────────────────

run(suite, "KMeans fit", "200x5 k=3", () =>
	new KMeans({ nClusters: 3, maxIter: 50 }).fit(clust200)
);
run(suite, "KMeans fit", "500x5 k=3", () =>
	new KMeans({ nClusters: 3, maxIter: 50 }).fit(clust500)
);
const km = new KMeans({ nClusters: 3, maxIter: 50 }).fit(clust200);
run(suite, "KMeans predict", "200x5", () => km.predict(clust200));

// ── DBSCAN ──────────────────────────────────────────────

run(suite, "DBSCAN fit", "200x5", () => new DBSCAN({ eps: 2.0, minSamples: 5 }).fit(clust200));
run(suite, "DBSCAN fit", "500x5", () => new DBSCAN({ eps: 2.0, minSamples: 5 }).fit(clust500));

// ── PCA ─────────────────────────────────────────────────

run(suite, "PCA fit", "200x5 k=2", () => new PCA({ nComponents: 2 }).fit(clust200));
run(suite, "PCA fit", "500x5 k=3", () => new PCA({ nComponents: 3 }).fit(clust500));
const pca = new PCA({ nComponents: 2 }).fit(clust200);
run(suite, "PCA transform", "200x5", () => pca.transform(clust200));

// ── TSNE ────────────────────────────────────────────────

run(suite, "TSNE fit", "200x5", () => new TSNE({ nComponents: 2, perplexity: 30 }).fit(clust200), {
	iterations: 3,
	warmup: 1,
});

footer(suite, "deepbox-ml.json");
